File: Organism.cpp
//
// Created by karol on 28.03.2025.
//

#include "include/Organism.h"
#include "include/World.h"

Organism::Organism(World* world) : world(world), strength(0), initiative(0), position(Point(0,0)), age(0) {
    // Constructor implementation
}

Organism::Organism(World* world, int strength, int initiative, Point position, string symbol) : world(world), strength(strength), initiative(initiative), position(position), age(0), symbol(symbol) {
    // Constructor implementation
}

Organism::Organism(World* world, int strength, int initiative, Point position, string symbol, int age) : world(world), strength(strength), initiative(initiative), position(position), age(age), symbol(symbol) {
    // Constructor implementation
}

Organism::~Organism() {
    // Destructor implementation
}

void Organism::kill(Organism& other) const {
    world->remove(other.getPosition());
}

int Organism::getStrength() const {
    return strength;
}

int Organism::getInitiative() const {
    return initiative;
}

int Organism::getAge() const {
    return age;
}

string Organism::getSymbol() const {
    return symbol;
}

void Organism::die() {
    //TODO
}

void Organism::action(char input) {
    action();
}

Point& Organism::getPosition() {
    return position;
}

void Organism::setPosition(Point pos) {
    position = pos;
}

void Organism::setAge(int age) {
    this->age = age;
}

void Organism::setStrength(int strength) {
    this->strength = strength;
}

File: game-of-life-java/src/Organism.java
import lib.Size;

import javax.swing.*;
import java.awt.*;
import java.io.Serializable;

public abstract class Organism implements Serializable {
    private static final long serialVersionUID = 1L;

    private int strength;
    protected int initiative;
    protected Point position;
    protected transient World world;
    protected int age;
    private boolean hasActed;

    public Organism(int strength, int initiative, Point position, World world, int age) {
        this.strength = strength;
        this.initiative = initiative;
        this.position = position;
        this.world = world;
        this.age = age;
    }

    public void setWorld(World world) {
        this.world = world;
    }

    public int getStrength() {
        return strength;
    }

    public void setStrength(int strength) {
        this.strength = strength;
    }

    protected void plantCollision(Organism plant, Organism animal) {
        //
    }

    public int getInitiative() {
        return initiative;
    }

    public Point getPosition() {
        return position;
    }

    public int getAge() {
        return age;
    }

    public boolean getHasActed() {
        return hasActed;
    }

    public void setHasActed(boolean hasActed) {
        this.hasActed = hasActed;
    }

    public void action() {

    }

    protected int escapeCollision(Organism other) {
        return 0;
    }

    public void action(String input) {
        action();
    }

    public void collision(Organism other) {
        // Domyślna implementacja, jeśli nie zostanie nadpisana w klasie potomnej
    }

    public ImageIcon getIcon() {
        return null;
    }
}
